File: src/algorithms/searching/binary/interpolation-search/interpolation-search.test.ts
import { describe, it, expect } from "vitest";

import { interpolationSearch } from "./sources/interpolation-search.ts?fn";

describe("interpolationSearch", () => {
  it("finds a value present in the array", () => {
    expect(interpolationSearch([2, 5, 8, 12, 16, 23, 38, 56, 72, 91], 23)).toBe(5);
  });

  it("returns -1 when the value is not in the array", () => {
    expect(interpolationSearch([2, 5, 8, 12, 16, 23, 38, 56, 72, 91], 50)).toBe(-1);
  });

  it("handles an empty array", () => {
    expect(interpolationSearch([], 5)).toBe(-1);
  });

  it("handles a single element array when found", () => {
    expect(interpolationSearch([42], 42)).toBe(0);
  });

  it("handles a single element array when not found", () => {
    expect(interpolationSearch([42], 10)).toBe(-1);
  });

  it("finds the first element in the array", () => {
    expect(interpolationSearch([2, 5, 8, 12, 16, 23, 38, 56, 72, 91], 2)).toBe(0);
  });

  it("finds the last element in the array", () => {
    expect(interpolationSearch([2, 5, 8, 12, 16, 23, 38, 56, 72, 91], 91)).toBe(9);
  });

  it("finds the middle element in the array", () => {
    expect(interpolationSearch([10, 20, 30, 40, 50], 30)).toBe(2);
  });

  it("returns -1 for a value smaller than all elements", () => {
    expect(interpolationSearch([5, 10, 15, 20], 1)).toBe(-1);
  });

  it("returns -1 for a value larger than all elements", () => {
    expect(interpolationSearch([5, 10, 15, 20], 100)).toBe(-1);
  });

  it("handles uniformly distributed data correctly", () => {
    const uniformArray = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100];
    expect(interpolationSearch(uniformArray, 70)).toBe(6);
  });

  it("handles duplicate values via division-by-zero guard", () => {
    expect(interpolationSearch([5, 5, 5, 5, 5], 5)).toBe(0);
  });

  it("returns -1 for a target not in a uniform-value array", () => {
    expect(interpolationSearch([5, 5, 5, 5, 5], 7)).toBe(-1);
  });
});

File: src/algorithms/searching/binary/interpolation-search/__tests__/InterpolationSearch_test.java
public class InterpolationSearch_test {
    public static void main(String[] args) {
        int[] standardArray = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};

        assert InterpolationSearch.interpolationSearch(standardArray, 23) == 5 : "should find value present";
        assert InterpolationSearch.interpolationSearch(standardArray, 50) == -1 : "should return -1 when not found";
        assert InterpolationSearch.interpolationSearch(new int[]{}, 5) == -1 : "should handle empty array";
        assert InterpolationSearch.interpolationSearch(new int[]{42}, 42) == 0 : "should find single element";
        assert InterpolationSearch.interpolationSearch(new int[]{42}, 10) == -1 : "should return -1 for single element not found";
        assert InterpolationSearch.interpolationSearch(standardArray, 2) == 0 : "should find first element";
        assert InterpolationSearch.interpolationSearch(standardArray, 91) == 9 : "should find last element";
        assert InterpolationSearch.interpolationSearch(new int[]{10, 20, 30, 40, 50}, 30) == 2 : "should find middle element";
        assert InterpolationSearch.interpolationSearch(new int[]{5, 10, 15, 20}, 1) == -1 : "should return -1 for smaller than all";
        assert InterpolationSearch.interpolationSearch(new int[]{5, 10, 15, 20}, 100) == -1 : "should return -1 for larger than all";
        assert InterpolationSearch.interpolationSearch(new int[]{10, 20, 30, 40, 50, 60, 70, 80, 90, 100}, 70) == 6 : "should handle uniformly distributed data";
        assert InterpolationSearch.interpolationSearch(new int[]{5, 5, 5, 5, 5}, 5) == 0 : "should handle duplicate values";
        assert InterpolationSearch.interpolationSearch(new int[]{5, 5, 5, 5, 5}, 7) == -1 : "should return -1 for target not in uniform array";

        System.out.println("All tests passed!");
    }
}
